### PR TITLE
Verify that no non deprecated SPI is removed

### PR DESCRIPTION
**Pull request by kokosing**
Verify that no non deprecated SPI is removed

This test verifies that SPI cannot be changed without a deprecation
period.

This partially solves the single version backward compatibility, the
other part is to ensure that engine still calls the deprecated API.

### CODE REVIEW

**Comment by kokosing** [comment]
Partially fixes https://github.com/trinodb/trino/issues/11667, the other part is not easily doable via tests. As we would need to have a plugins in previous version. I guess it requires the code review process to make sure that the old API is still used.

File: core/trino-spi/src/test/java/io/trino/spi/TestSingleVersionBackwardCompatibility.java
Comment (martint) [inline core/trino-spi/src/test/java/io/trino/spi/TestSingleVersionBackwardCompatibility.java:62]: This will cause tests to fail in offline mode, or in any environment that doesn't have direct access to maven central. In some organizations, all access is performed through a proxy or a local maven repository.

**Review by martint — CHANGES_REQUESTED** [review]

File: core/trino-spi/src/test/java/io/trino/spi/TestSingleVersionBackwardCompatibility.java
Comment (kokosing) [reply core/trino-spi/src/test/java/io/trino/spi/TestSingleVersionBackwardCompatibility.java:62]: So it looks like I need to download artifacts via maven, so it should use local maven repository in most of the cases.

Comment (electrum) [reply core/trino-spi/src/test/java/io/trino/spi/TestSingleVersionBackwardCompatibility.java:62]: You could probably use https://maven.apache.org/plugins/maven-dependency-plugin/copy-mojo.html#dependency-copy

Comment (findepi) [reply core/trino-spi/src/test/java/io/trino/spi/TestSingleVersionBackwardCompatibility.java:62]: I think it's reasonable to require internet access during tests.
For example, docker-based tests won't pass if it is not possible to download the container images.

cc @nineinchnick

File: core/trino-spi/pom.xml
Comment (findepi) [inline core/trino-spi/pom.xml:114]: be specific which file(s) should be filtered

File: core/trino-spi/src/test/java/io/trino/spi/TestSingleVersionBackwardCompatibility.java
Comment (findepi) [reply core/trino-spi/src/test/java/io/trino/spi/TestSingleVersionBackwardCompatibility.java:62]: I agree using local maven repo would be better though. Eg it would cover for retries required to cope with central's instability.

This can help

```xml
        <plugin>
                <groupId>org.apache.maven.plugins</groupId>
                <artifactId>maven-dependency-plugin</artifactId>
                <executions>
                    <execution>
                        <id>copy</id>
                        <phase>package</phase>
                        <goals>
                            <goal>copy</goal>
                        </goals>
                        <configuration>
                            <skip>false</skip>
                            <artifactItems>
                                <artifactItem>
                                    <groupId>io.trino</groupId>
                                    <artifactId>trino-cli</artifactId>
                                    <version>${dep.trino.version}</version>
                                    <classifier>executable</classifier>
                                    <type>jar</type>
                                    <outputDirectory>${project.build.directory}</outputDirectory>
                                </artifactItem>
```

File: core/trino-spi/src/test/java/io/trino/spi/TestSingleVersionBackwardCompatibility.java
Comment (findepi) [inline core/trino-spi/src/test/java/io/trino/spi/TestSingleVersionBackwardCompatibility.java:63]: SPI has more deps than just slice. Why is the Slice special? Document

File: core/trino-spi/src/test/java/io/trino/spi/TestSingleVersionBackwardCompatibility.java
Comment (findepi) [inline core/trino-spi/src/test/java/io/trino/spi/TestSingleVersionBackwardCompatibility.java:72]: url1 -> inputStream

File: core/trino-spi/src/test/java/io/trino/spi/TestSingleVersionBackwardCompatibility.java
Comment (findepi) [inline core/trino-spi/src/test/java/io/trino/spi/TestSingleVersionBackwardCompatibility.java:48]: will we need an scape hatch here?

File: core/trino-spi/src/test/java/io/trino/spi/TestSingleVersionBackwardCompatibility.java
Comment (findepi) [inline core/trino-spi/src/test/java/io/trino/spi/TestSingleVersionBackwardCompatibility.java:61]: Are we assuming that Slice version doesn't change?

File: core/trino-spi/src/test/java/io/trino/spi/TestSingleVersionBackwardCompatibility.java
Comment (martint) [reply core/trino-spi/src/test/java/io/trino/spi/TestSingleVersionBackwardCompatibility.java:62]: > docker-based tests won't pass if it is not possible to download the container images.

You can always pre-download them and run the tests offline.

**Comment by martint** [comment]
> This test verifies that SPI cannot be changed without a deprecation
period.

Sometimes that is not possible, and we have to break backward compatibility. This change will entirely prevent that and remove that ability.

File: core/trino-spi/src/test/java/io/trino/spi/TestSingleVersionBackwardCompatibility.java
Comment (kokosing) [reply core/trino-spi/src/test/java/io/trino/spi/TestSingleVersionBackwardCompatibility.java]: Slice is special as loading SPI classes was loading Slice classes. Others were not loaded.

File: core/trino-spi/src/test/java/io/trino/spi/TestSingleVersionBackwardCompatibility.java
Comment (kokosing) [reply core/trino-spi/src/test/java/io/trino/spi/TestSingleVersionBackwardCompatibility.java]: It can change, however I assume that loading slice classes for SPI for previous version will still work. Otherwise it sounds like yet another backward incompatible change.

**Review by kokosing — COMMENTED** [review]
AC

File: core/trino-spi/src/test/java/io/trino/spi/TestSingleVersionBackwardCompatibility.java
Comment (findepi) [reply core/trino-spi/src/test/java/io/trino/spi/TestSingleVersionBackwardCompatibility.java]: I am not sure i agree, but please capture your reasoning as a code comment

**Comment by kokosing** [comment]
> Sometimes that is not possible, and we have to break backward compatibility. This change will entirely prevent that and remove that ability.

@martint Notice that this test captured backward incompatibilities that were introduced in current ongoing version. So test does not prevent them, it makes them visible. Now we could potentially add them to release notes.

File: core/trino-spi/src/test/java/io/trino/spi/TestSingleVersionBackwardCompatibility.java
Comment (findepi) [inline core/trino-spi/src/test/java/io/trino/spi/TestSingleVersionBackwardCompatibility.java:49]: exclusions (BACKWARD_INCOMPATIBLE_CHANGES) should be verified for being part of previous SPI

eg a test should fail when i add `public void foo()` to BACKWARD_INCOMPATIBLE_CHANGES, unless such method existed in the previous SPI AND it no longer exists in the new SPI.

Comment (findepi) [reply core/trino-spi/src/test/java/io/trino/spi/TestSingleVersionBackwardCompatibility.java:49]: Alternatively, we could eg have `BACKWARD_INCOMPATIBLE_CHANGES` as a map `version -> changes`
and we would just look at the entry for current version.
Of course, the point is not to keep track of all the breaking changes ever, but rather not to fail right after a release.
Then next person adding something new to the list would remove old (obsolete) entries.

Comment (kokosing) [reply core/trino-spi/src/test/java/io/trino/spi/TestSingleVersionBackwardCompatibility.java:49]: I no longer get a version from the pom to java. So in tests I have no notion of a version. Bringing back the version in the test just for this sounds like an overkill to me.

Maybe I can simply add a comment that:

```
// when updating  BACKWARD_INCOMPATIBLE_CHANGES please check if you can remove obsolete entries
```
?

Comment (findepi) [reply core/trino-spi/src/test/java/io/trino/spi/TestSingleVersionBackwardCompatibility.java:49]: > Bringing back the version in the test just for this sounds like an overkill to me.

It's not

BTW i am making Java aware of SPI version in https://github.com/trinodb/trino/pull/11774

**Comment by kokosing** [comment]
@findepi AC

@martint PTAL, I hope I have addressed your concerns.

File: core/trino-spi/src/test/java/io/trino/spi/TestSingleVersionBackwardCompatibility.java
Comment (martint) [inline core/trino-spi/src/test/java/io/trino/spi/TestSingleVersionBackwardCompatibility.java:62]: This is generally the case, but might not always be. For instance, if something happens during the release process, we might need to "burn" a release (e.g., https://trino.io/docs/current/release/release-349.html)

Comment (findepi) [reply core/trino-spi/src/test/java/io/trino/spi/TestSingleVersionBackwardCompatibility.java:62]: we will have to change this code line when doing that. perhaps add an `if`.

**Comment by kokosing** [comment]
@martint AC, PTAL

File: core/trino-spi/pom.xml
Comment (findepi) [inline core/trino-spi/pom.xml:18]: This should be auto-derived IMO. Why hard-coded? it adds maintenance overhead.

Note that when the test fails people will be steered first towards maintaining compatibility with an old version, rather than fix the version number here.

File: core/trino-spi/src/test/java/io/trino/spi/TestSpiBackwardCompatibility.java
Comment (findepi) [inline core/trino-spi/src/test/java/io/trino/spi/TestSpiBackwardCompatibility.java:115]: Should we consider nested classes too?

File: core/trino-spi/src/test/java/io/trino/spi/TestSpiBackwardCompatibility.java
Comment (findepi) [inline core/trino-spi/src/test/java/io/trino/spi/TestSpiBackwardCompatibility.java:127]: This is recursive (higher up in class hierarchy), maybe worth commenting about this.

Also, you returning Object-defined methods

```
public final void java.lang.Object.wait(long,int) throws java.lang.InterruptedException
public final void java.lang.Object.wait() throws java.lang.InterruptedException
public final native void java.lang.Object.wait(long) throws java.lang.InterruptedException
public boolean java.lang.Object.equals(java.lang.Object)
public java.lang.String java.lang.Object.toString()
public native int java.lang.Object.hashCode()
public final native java.lang.Class<?> java.lang.Object.getClass()
public final native void java.lang.Object.notify()
public final native void java.lang.Object.notifyAll()
```

they will not cause a change (unless last SPI classes is removed), but still maybe worth documenting?

File: core/trino-spi/src/test/java/io/trino/spi/TestSpiBackwardCompatibility.java
Comment (findepi) [inline core/trino-spi/src/test/java/io/trino/spi/TestSpiBackwardCompatibility.java:51]: This is an example entry which makes adding new entries easier, but this example entry should be removed already (per comment above + 376 is out). Let's make adding new entries easier in some more durable manner, perhaps as a commented-out copy-paste-ready example.

File: core/trino-spi/src/test/java/io/trino/spi/TestSpiBackwardCompatibility.java
Comment (findepi) [inline core/trino-spi/src/test/java/io/trino/spi/TestSpiBackwardCompatibility.java:92]: cwd may or may not be module's dir.

- does it work when run from IDE (with your default settings)?
- does it work when run as `mvn test -pl :trino-spi`?
- does it work when run as `cd core/trino-spi && mvn test`?

File: core/trino-spi/pom.xml
Comment (kokosing) [reply core/trino-spi/pom.xml:18]: This is about addressing this comment: https://github.com/trinodb/trino/pull/11722#discussion_r845510571

> Note that when the test fails people will be steered first towards maintaining compatibility with an old version, rather than fix the version number here.

Maybe it is a good thing. Also I can add some comment here to make it more clear what is the goal here. Like `We try to be backward compatible with at least last released version`. WDYT?

File: core/trino-spi/src/test/java/io/trino/spi/TestSpiBackwardCompatibility.java
Comment (kokosing) [reply core/trino-spi/src/test/java/io/trino/spi/TestSpiBackwardCompatibility.java]: I am not sure. 375-SNAPSHOT is almost compatible with 375 which is a good thing.

Also since now version is hardcoded, I no longer think that the key (`"375"`) is needed here.

File: core/trino-spi/src/test/java/io/trino/spi/TestSpiBackwardCompatibility.java
Comment (kokosing) [reply core/trino-spi/src/test/java/io/trino/spi/TestSpiBackwardCompatibility.java:70]: Yes. Yes. Yes.

File: core/trino-spi/src/test/java/io/trino/spi/TestSpiBackwardCompatibility.java
Comment (kokosing) [reply core/trino-spi/src/test/java/io/trino/spi/TestSpiBackwardCompatibility.java]: I changed it to return only declared entities and skip object and enum methods. Thanks.

Previously, used `getMethods` and `getFields` as it returns only public things. However it is easier public things than hierarchical ones (or maybe it as same level of complexity...)

**Review by kokosing — COMMENTED** [review]
AC

File: core/trino-spi/src/test/java/io/trino/spi/TestSingleVersionBackwardCompatibility.java
Comment (findepi) [reply core/trino-spi/src/test/java/io/trino/spi/TestSingleVersionBackwardCompatibility.java]: note also that we would have same problem with JDBC and server compatibility testing. 
A rare case of a burned release would need to be handled explicitly.

File: core/trino-spi/pom.xml
Comment (findepi) [reply core/trino-spi/pom.xml:18]: > Maybe it is a good thing.

This isn't what was intended and is more than we wanted.

>  add some comment here to make it more clear 

the problem is -- the comment here won't be visible; the test failure will

> This is about addressing this comment: https://github.com/trinodb/trino/pull/11722#discussion_r845510571

i replied there

File: core/trino-spi/src/test/java/io/trino/spi/TestSpiBackwardCompatibility.java
Comment (findepi) [inline core/trino-spi/src/test/java/io/trino/spi/TestSpiBackwardCompatibility.java:123]: is this reachable?

(i think it's not)

File: core/trino-spi/src/test/java/io/trino/spi/TestSpiBackwardCompatibility.java
Comment (findepi) [inline core/trino-spi/src/test/java/io/trino/spi/TestSpiBackwardCompatibility.java:130]: it would be more readable if `entities.add(` was last line in the loop body, and deprecation handling was done as all other skipping (`if (condition) continue`)

File: core/trino-spi/src/test/java/io/trino/spi/TestSpiBackwardCompatibility.java
Comment (findepi) [inline core/trino-spi/src/test/java/io/trino/spi/TestSpiBackwardCompatibility.java:126]: is this reachable?

(i think it's not)

File: core/trino-spi/src/test/java/io/trino/spi/TestSpiBackwardCompatibility.java
Comment (findepi) [inline core/trino-spi/src/test/java/io/trino/spi/TestSpiBackwardCompatibility.java:138]:
```suggestion
            if (!includeDeprecated && field.isAnnotationPresent(Deprecated.class)) {
                continue;
            }
            entities.add("Field: " + field.toGenericString());
```

File: core/trino-spi/src/test/java/io/trino/spi/TestSpiBackwardCompatibility.java
Comment (findepi) [reply core/trino-spi/src/test/java/io/trino/spi/TestSpiBackwardCompatibility.java]: The purpose of key-ing on version was to enforce that backwards incompatible change declaration is not implicitly carried over to the next release.

Let's imagine

- v 100: we have a public method foo
- v 101: we remove the method foo and declare this as a backward incompatible change
- v 102: we reintroduce method foo
- v 103-SNAPSHOT: we remove method foo -- we want the test to fail, as this is a _new_ backward incompatible change; for it to fail, we need to ensure the 101's suppression is not carried forwards implicitly

File: core/trino-spi/src/test/java/io/trino/spi/TestSpiBackwardCompatibility.java
Comment (findepi) [reply core/trino-spi/src/test/java/io/trino/spi/TestSpiBackwardCompatibility.java:71]: Awesome! 

Curious, how does it work then? i remember we had some problem with similar code here
https://github.com/trinodb/trino/blob/7cac412b9e69acf39b2e7c2111767f16998619d7/testing/trino-benchto-benchmarks/src/test/java/io/trino/sql/planner/AbstractCostBasedPlanTest.java#L131-L137

Comment (kokosing) [reply core/trino-spi/src/test/java/io/trino/spi/TestSpiBackwardCompatibility.java:71]: I recall that too. Maybe something changed in maven?

File: core/trino-spi/pom.xml
Comment (findepi) [reply core/trino-spi/pom.xml:158]: see now, it's from `build-helper-maven-plugin`.

File: core/trino-spi/src/test/java/io/trino/spi/TestSpiBackwardCompatibility.java
Comment (findepi) [inline core/trino-spi/src/test/java/io/trino/spi/TestSpiBackwardCompatibility.java:71]: `BACKWARD_INCOMPATIBLE_CHANGES.get...(currentVersion)` -- "the backwards incompatible changes introduced in version (current) are ..."

File: core/trino-spi/pom.xml
Comment (kokosing) [reply core/trino-spi/pom.xml:138]: See https://www.mojohaus.org/build-helper-maven-plugin/released-version-mojo.html

> Resolve the latest released version of this project. 

I am not sure what it does internally.

**Comment by kokosing** [comment]
AC

@findepi Thank you for the review! ❤️

**Review by findepi — APPROVED** [review]

**Comment by findepi** [comment]
@kokosing thank you for your patience with my review comments.
This is an awesome addition!

I am gonna merge this tomorrow, unless anyone has any more comments.

cc @trinodb/maintainers

**Comment by kokosing** [comment]
> thank you for your patience with my review comments.

@findepi This was a pleasure. Always! I appreciate all your comments, they improved this change a lot!

**Comment by kokosing** [comment]
Thank you!

**Pull request merged**
